### ISSUE TITLE
Zooming Container

### ISSUE THREAD

**Issue by Kethku** [linebender/druid#1889]
I've been working on a tree based task management app in druid that has a component which allows placing and moving of widgets on an infinite canvas. It looks something like this: 
![image](https://user-images.githubusercontent.com/3323631/128062754-38ba0a8c-9e88-40e9-9bc5-ad25068820b8.png)

This works great modulo some performance issues, but I can't figure out how to implement zooming of the canvas. I've seen some requests and suggestions that such a thing should exist, but I haven't seen any suggestions for how it should be implemented. I'm guessing it would require modifying of events before passing them to child widgets and some trickery with the paint context.

I'm interested in working on such features. Does anybody here have any suggestions for how to get started on such a thing?

**Comment by johannesvollmer** [comment]
I'm also missing a generic single-child container that has a single `Affine` transform as main property. Like CSS' `transform` property. 

Transforming events could maybe be as simple as computing the inverse of the affine, and then feeding the event position through that inverse transform?

**Comment by jneem** [comment]
This is something that has come up (most recently in #1799), and so far the decision has been that we don't want to support transforms of arbitrary widget trees, exactly because of the issue with layout units changing for the children. For example, built-in widgets try to align themselves to the pixel grid, and that will behave strangely with zooming.

The toolkits I'm most familiar with (gtk and qt) also don't support transformations of widgets. Instead, their canvases have children which are not widgets, and which *do* support arbitrary transformations. Would something like that work for you?

**Comment by Kethku** [comment]
I'm not sure that would work. Both @johannesvollmer and I are trying to build something like a node based graph editor. I personally am trying to build something like https://reactflow.dev/ on top of druid. Doing so requires true widgets to have a good experience. It would be a bummer to reimplement text editors/sliders/buttons etc from scratch just to implement zooming

**Comment by Kethku** [comment]
Just to follow up on #1799 the suggestion there and above is for a full affine transformation which to my knowledge includes skew. I think thats likely overkill. Maybe the concerns brought up there can be partially be addressed by only allowing for scale and translation and not the skew operations. Not sure

**Comment by Kethku** [comment]
That said, the referenced PR gives a pattern for how I could implement this myself, so I'm not exactly blocked. Just figured I'd make the request since it would be nice if this just worked using off the shelf components rather than implementing it myself.

**Comment by johannesvollmer** [comment]
For my specific use case, scale and offset would be enough. Skew would not be important to me. 

If I had to use a canvas, I would not bother using druid, as virtually all interactions happen within the scaled widget. I'm pretty sure it would be easier to get scaling right than to re-implement all widgets for a canvas. 

I know scaling is not important for most apps, as they could use canvas, but for node editors, there's really no other option as far as I know

**Comment by jneem** [comment]
Have you thought about how to handle pixel alignment? Would you want a flag that disables it for a subtree? Or some way to "reach through" all the transformations and find where the pixel grid is?

**Comment by johannesvollmer** [comment]
If we use only scaling and translation, passing the pixel grid down to the children might be the cleanest solution. I'm not sure how druid works yet though.

**Comment by Kethku** [comment]
I agree passing the grid down seems like the correct move. Obviously that would preclude using rotation and skew, but like mentioned above, I think there is value in just adding the ability to zoom the canvas

**Comment by johannesvollmer** [comment]
If we have a container that does both scaling and translation, it could save the state as f32, but before passing it on to the children, perform a rounding operation such that the transformation keeps the grid? Is that enough or do the children need more?
